File: src/token.zig
const std = @import("std");
const mem = std.mem;

pub const TokenType = enum {
    illegal,
    eof,
    // Identifiers
    read_ident,
    ident,
    // Types
    int,
    // Operators
    assign,
    plus,
    minus,
    bang,
    asterisk,
    slash,
    less_than,
    greater_than,
    equal_equal,
    not_equal,
    // Syntax
    comma,
    semicolon,
    l_paren,
    r_paren,
    l_brace,
    r_brace,
    // Keywords
    let,
    _if,
    _else,
    function,
    _return,
    _true,
    _false,
};

pub const Token = struct {
    t: TokenType,
    literal: []const u8,

    pub fn init(t: TokenType, literal: []const u8) Token {
        return Token{
            .t = t,
            .literal = literal,
        };
    }
};

pub fn is_letter(ch: u8) bool {
    return ('a' <= ch and ch <= 'z') or ('A' <= ch and ch <= 'Z') or ch == '_';
}

pub fn is_digit(ch: u8) bool {
    return '0' <= ch and ch <= '9';
}

pub fn lookup_keywords(keyword: []const u8) TokenType {
    if (mem.eql(u8, keyword, "let")) {
        return .let;
    } else if (mem.eql(u8, keyword, "fn")) {
        return .function;
    } else if (mem.eql(u8, keyword, "return")) {
        return ._return;
    } else if (mem.eql(u8, keyword, "if")) {
        return ._if;
    } else if (mem.eql(u8, keyword, "else")) {
        return ._else;
    } else if (mem.eql(u8, keyword, "true")) {
        return ._true;
    } else if (mem.eql(u8, keyword, "false")) {
        return ._false;
    } else {
        return .ident;
    }
}

File: lib/token.ex
defmodule Token do
  @enforce_keys [:type, :literal]
  defstruct [:type, :literal]

  @keywords %{
    "fn" => :fn,
    "let" => :let,
    "true" => true,
    "false" => false,
    "if" => :if,
    "else" => :else,
    "return" => :return
  }

  @types %{
    illegal: "ILLEGAL",
    eof: "EOF",
    # identifiers and literals
    ident: "IDENT",
    # 123
    int: "INT",
    # operators
    assign: "=",
    plus: "+",
    minus: "-",
    asterisk: "*",
    slash: "/",
    bang: "!",
    equal_equal: "==",
    not_equal: "!=",
    greater_than: ">",
    less_than: "<",
    # delimiters
    comma: ",",
    semicolon: ",",
    lparen: "(",
    rparen: ")",
    lbrace: "{",
    rbrace: "}",
    fn: "FUNCTION",
    let: "LET",
    true: "TRUE",
    false: "FALSE",
    if: "IF",
    else: "ELSE",
    return: "RETURN"
  }

  @spec new(type :: atom(), literal :: String.t()) :: %Token{} | no_return()
  def new(type, literal) when is_atom(type) and is_binary(literal) do
    if Map.has_key?(@types, type) do
      %__MODULE__{type: type, literal: literal}
    else
      raise "Unsupported token type: #{inspect(type)}"
    end
  end

  @spec lookup_identifier(ident :: String.t()) :: String.t() | nil 
  def lookup_identifier(ident) do
    Map.get(@keywords, ident, :ident)
  end
end
